File: src/main/java/art/arcane/racbot/discord/CommandRegistrar.java
package art.arcane.racbot.discord;

import art.arcane.racbot.config.BotConfig;
import art.arcane.racbot.config.CommandRegistrationMode;
import java.util.ArrayList;
import java.util.List;
import net.dv8tion.jda.api.JDA;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.interactions.commands.build.CommandData;
import net.dv8tion.jda.api.interactions.commands.build.Commands;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class CommandRegistrar {
  private static final Logger LOGGER = LoggerFactory.getLogger(CommandRegistrar.class);

  private final BotConfig config;

  public CommandRegistrar(BotConfig config) {
    this.config = config;
  }

  public void register(JDA jda) {
    List<CommandData> commands = buildCommands();
    if (commands.isEmpty()) {
      LOGGER.warn("No commands enabled. Skipping slash command registration.");
      return;
    }

    if (config.bot().commandRegistrationMode() == CommandRegistrationMode.GUILD
        && config.bot().devGuildId() != null) {
      Guild guild = jda.getGuildById(config.bot().devGuildId());
      if (guild != null) {
        guild.updateCommands().addCommands(commands).queue();
        LOGGER.info("Registered {} commands in dev guild {}", commands.size(), guild.getId());
        return;
      }
      LOGGER.warn(
          "Configured dev guild {} not found, falling back to global command registration.",
          config.bot().devGuildId());
    }

    jda.updateCommands().addCommands(commands).queue();
    LOGGER.info("Registered {} commands globally.", commands.size());
  }

  private List<CommandData> buildCommands() {
    List<CommandData> commands = new ArrayList<>();

    if (config.features().pingEnabled()) {
      commands.add(Commands.slash("ping", "Check bot latency"));
    }

    return commands;
  }
}

File: lib/src/discord/command_registrar.dart
import 'package:nyxx/nyxx.dart';
import 'package:nyxx_commands/nyxx_commands.dart';
import 'package:racbot_nyxx/src/config/bot_config.dart';
import 'package:racbot_nyxx/src/config/command_registration_mode.dart';
import 'package:racbot_nyxx/src/discord/ping_command_module.dart';
import 'package:racbot_nyxx/src/util/app_logger.dart';

class CommandRegistrar {
  final AppLogger logger;

  const CommandRegistrar({required this.logger});

  Future<void> register({
    required NyxxGateway client,
    required CommandsPlugin commandsPlugin,
    required PingCommandModule pingCommandModule,
    required BotConfig config,
  }) async {
    if (!config.features.pingEnabled) {
      logger.warning(
        'No commands enabled. Skipping slash command registration.',
      );
      return;
    }

    if (!pingCommandModule.pingCommandAdded) {
      logger.warning(
        'Ping command is enabled in config but not attached to nyxx_commands yet.',
      );
      return;
    }

    List<ApplicationCommandBuilder> builders = <ApplicationCommandBuilder>[
      ApplicationCommandBuilder.chatInput(
        name: 'ping',
        description: 'Check bot latency',
        options: <CommandOptionBuilder>[],
      ),
    ];

    if (config.bot.commandRegistrationMode == CommandRegistrationMode.guild &&
        config.bot.devGuildId != null) {
      Snowflake guildId = Snowflake(config.bot.devGuildId!);
      if (client.guilds.cache.containsKey(guildId)) {
        List<ApplicationCommand> registered = await client
            .guilds[guildId]
            .commands
            .bulkOverride(builders);

        commandsPlugin.registeredCommands.clear();
        commandsPlugin.registeredCommands.addAll(registered);

        logger.info(
          'Registered ${registered.length} commands in dev guild ${guildId.value}',
        );
        return;
      }

      logger.warning(
        'Configured dev guild ${config.bot.devGuildId} not found, falling back to global command registration.',
      );
    }

    List<ApplicationCommand> registered = await client.commands.bulkOverride(
      builders,
    );
    commandsPlugin.registeredCommands.clear();
    commandsPlugin.registeredCommands.addAll(registered);
    logger.info('Registered ${registered.length} commands globally.');
  }
}
